src/dst/src/Distributed.cc

```diff
@@ -0,0 +1,201 @@
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE REGENTS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dst/Distributed.h"
+
+#include <boost/asio.hpp>
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/system/system_error.hpp>
+#include <vector>
+
+#include "LoadBalancer.h"
+#include "Worker.h"
+#include "dst/JobCallBack.h"
+#include "dst/JobMessage.h"
+#include "sta/StaMain.hh"
+#include "utl/Logger.h"
+namespace dst {
+const int MAX_TRIES = 5;
+}
+
+using namespace dst;
+
+namespace sta {
+// Tcl files encoded into strings.
+extern const char* dst_tcl_inits[];
+}  // namespace sta
+
+extern "C" {
+extern int Dst_Init(Tcl_Interp* interp);
+}
+
+Distributed::Distributed() : logger_(nullptr)
+{
+}
+
+Distributed::~Distributed()
+{
+  for (auto cb : callbacks_)
+    delete cb;
+  callbacks_.clear();
+}
+
+void Distributed::init(Tcl_Interp* tcl_interp, utl::Logger* logger)
+{
+  logger_ = logger;
+  // Define swig TCL commands.
+  Dst_Init(tcl_interp);
+  sta::evalTclInit(tcl_interp, sta::dst_tcl_inits);
+}
+
+void Distributed::runWorker(unsigned short port)
+{
+  try {
+    asio::io_service io_service;
+    Worker worker(io_service, this, logger_, port);
+    io_service.run();
+  } catch (std::exception& e) {
+    logger_->error(utl::DST, 1, "Worker server error: {}", e.what());
+  }
+}
+
+void Distributed::runLoadBalancer(unsigned short port)
+{
+  try {
+    asio::io_service io_service;
+    LoadBalancer balancer(io_service, logger_, port);
+    for (auto worker : workers_)
+      balancer.addWorker(worker.ip, worker.port, 10);
+    io_service.run();
+  } catch (std::exception& e) {
+    logger_->error(utl::DST, 9, "LoadBalancer error: {}", e.what());
+  }
+}
+
+void Distributed::addWorkerAddress(const char* address, unsigned short port)
+{
+  workers_.push_back(EndPoint(address, port));
+}
+
+bool sendMsg(dst::socket& sock, const std::string& msg, std::string& errorMsg)
+{
+  int tries = 0;
+  while (tries++ < MAX_TRIES) {
+    boost::system::error_code error;
+    write(sock, asio::buffer(msg), error);
+    if (!error) {
+      errorMsg.clear();
+      return true;
+    } else
+      errorMsg = error.message();
+  }
+  return false;
```

Comment:
This is fine for now, but you might want to add a TODO on using exponential backoff. Linear backoff has a propensity to cause massive traffic spikes if there's a short disconnection which can degrade the system even further. This is why it's better to use an existing protocol like gRPC or HTTP which has backoff semantics better defined in client libraries.